ledger/eval/eval.go

```diff
@@ -1611,20 +1616,68 @@ func (eval *BlockEvaluator) generateKnockOfflineAccountsList() {
 	if !eval.generate {
 		return
 	}
-	current := eval.Round()
 
+	current := eval.Round()
 	maxExpirations := eval.proto.MaxProposedExpiredOnlineAccounts
 	maxSuspensions := eval.proto.Payouts.MaxMarkAbsent
 
 	updates := &eval.block.ParticipationUpdates
 
-	ch := activeChallenge(&eval.proto, uint64(eval.Round()), eval.state)
+	ch := activeChallenge(&eval.proto, uint64(current), eval.state)
+
+	// Make a set of candidate addresses to check for absentee status.
+	type candidateData struct {
+		VoteLastValid         basics.Round
+		VoteID                crypto.OneTimeSignatureVerifier
+		Status                basics.Status
+		LastProposed          basics.Round
+		LastHeartbeat         basics.Round
+		MicroAlgosWithRewards basics.MicroAlgos
+		IncentiveEligible     bool // currently unused below, but may be needed in the future
+	}
+	absentCandidates := make(map[basics.Address]candidateData)
+
+	// First, ask the ledger for the top N online accounts, with their latest
+	// online account data, current up to the previous round.
+	if maxSuspensions > 0 {
+		knockOfflineCandidates, err := eval.state.knockOfflineCandidates()
```

Comment:
since this depends on a node state (not chain state) it could produce different results on different nodes - one can return actual voters and other voters for previous period. In this case network would not agree on block.
Even with "if a current period voters not ready then return empty" rule it is the same story, some might have it ready.
To mitigate consider this:
1. move the invocation of `generateKnockOfflineAccountsList` after the `stateProofVotersAndTotal` in `endOfBlock`
2. always aim for the current interval, and return empty if not ready.
This would allows us to ensure that a current interval voters are ready, and all nodes use the same input data.

Upd: I realized it is called only in mode=generate, so probably does not matter, but maybe makes sense to comment/document it is expected